testsuite/src/main/java/io/netty/testsuite/transport/socket/CompositeBufferGatheringWriteTest.java

```diff
@@ -282,6 +378,139 @@ public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+                            ctx.write(contents.readSplit(soSndBuf - 100));
+
+                            // Build and write CompositeByteBuf
+                            CompositeBuffer compositeBuffer = CompositeBuffer.compose(
+                                    ctx.bufferAllocator(),
+                                    contents.readSplit(50).send(),
+                                    contents.readSplit(200).send());
+                            ctx.write(compositeBuffer);
+
+                            // Write a single buffer that is smaller than the second component of the
+                            // CompositeBuffer above but small enough to fit in the remaining space allowed by the
+                            // soSndBuf amount.
+                            ctx.write(contents.readSplit(50));
+
+                            // Write the remainder of the content
+                            ctx.writeAndFlush(contents).addListener(ctx, ChannelFutureListeners.CLOSE);
+                        }
+
+                        @Override
+                        public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+                            // IOException is fine as it will also close the channel and may just be a connection reset.
+                            if (!(cause instanceof IOException)) {
+                                clientReceived.set(cause);
+                                latch.countDown();
+                            }
+                        }
+                    });
+                }
+            });
+            cb.handler(new ChannelInitializer<Channel>() {
+                @Override
+                protected void initChannel(Channel ch) throws Exception {
+                    ch.pipeline().addLast(new ChannelHandler() {
+                        private Buffer aggregator;
+                        @Override
+                        public void handlerAdded(ChannelHandlerContext ctx) {
+                            aggregator = ctx.bufferAllocator().allocate(expectedContent.readableBytes());
+                        }
+
+                        @Override
+                        public void channelRead(ChannelHandlerContext ctx, Object msg) {
+                            if (msg instanceof Buffer) {
+                                try (Buffer buf = (Buffer) msg) {
+                                    aggregator.writeBytes(buf);
+                                }
+                            }
+                        }
+
+                        @Override
+                        public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+                            // IOException is fine as it will also close the channel and may just be a connection reset.
+                            if (!(cause instanceof IOException)) {
+                                clientReceived.set(cause);
+                                latch.countDown();
```

Comment:
should we also close the aggregator here ?

Reply:
Does `channelInactive` not take care of it regardless?